### PR TITLE
Make perforation strategies time/state/change (plugs/straddles) aware

### LINKED ISSUE #163: How to model straddling and plugging in wells
Currently, wells are modeled as having single connections to the reservoir. We do not take into account any changes in the connections of a well to the reservoir by means of straddles and/or plugs.

This will make it hard/impossible to model jumps in water cut development.

How should we handle this?

### PR DESCRIPTION
**Pull request by wouterjdb**
This pull request improves the perforation strategies code and includes the following:

- Refactor the perforation strategies code
- Opening/closing individual connections in time
- A new perforation strategy that basis the number of connection on historic plugs/straddles. This should allow us to model discrete steps in, for example water cut, when a connection is straddled/plugged with a minimal number of connections to a FlowNet.

---

### Contributor checklist

- [x] :tada: This PR closes #163
- [x] :scroll: I have broken down my PR into the following tasks:
   - [x] Refactor code
   - [x] Create new strategy
   - [x] Implement changes through time for connections in schedule generation
   - [x] Fix black bug
   - [x] Fix config
   - [x] Add `WELOPEN` to `ecl2df` (https://github.com/equinor/ecl2df/pull/186)
   - [x] Implement reading of compdat data via ecl2df
   - [x] Support `COMPDAT` and `WELOPEN` in the same date (https://github.com/equinor/ecl2df/pull/189)
   - [x] Fix NAN in `WCONHIST` / `WCONINJH`
   - [x] Finalize testing
- [x] :robot: I have added tests, or extended existing tests, to cover any new features or bugs fixed in this PR.
- [x] :book: I have considered adding a new entry in `CHANGELOG.md`.
- [x] :books: I have considered updating the documentation.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@ This project adheres to [Semantic Versioning](https://semver.org/).
 ## [Unreleased]
 
 ### Added
+- [#197](https://github.com/equinor/flownet/pull/197) Added opening/closing of connects based on straddling/plugging/perforations through time. This PR also adds a new perforation strategy `multiple_based_on_workovers` which models the well connections with as few connections as possible taking into account opening/closing of groups of connections.
```
Comment (olelod) [inline CHANGELOG.md:8]:
```suggestion
- [#197](https://github.com/equinor/flownet/pull/197) Added opening/closing of connections based on straddling/plugging/perforations through time. This PR also adds a new perforation strategy `multiple_based_on_workovers` which models the well connections with as few connections as possible taking into account opening/closing of groups of connections. 
```

File: src/flownet/network_model/_network_model.py [hunk 2]
```diff
@@ -27,6 +28,8 @@ def __init__(
             df_entity_connections: A dataframe containing information about the individual one
                                    dimensional models. Columns required are xstart, ystart, zstart,
                                    xend, yend, zend, start_entity and end_entity
+            df_well_connections: A dataframe containing information about the (individual) connections of a
+                                 well, and its open or close state throughout time.
```
Comment (olelod) [inline src/flownet/network_model/_network_model.py:32]:
```suggestion
                                 well, and its open or closed state throughout time.
```

File: src/flownet/data/perforation_strategy.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+        .assign(OPEN=True)
+        .merge(df_first_dates, on="WELL_NAME")
+    )
+    return result
+
+
+def multiple(df: pd.DataFrame) -> pd.DataFrame:
+    """
+    This strategy creates multiple connections per well, as many as there is data available. Connections that
+    repeatedly have the same state through time are reduced to only only having records for state changes.
```
Comment (olelod) [inline src/flownet/data/perforation_strategy.py:74]:
```suggestion
    repeatedly have the same state through time are reduced to only having records for state changes.
```

File: src/flownet/data/perforation_strategy.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+                & (
+                    (df_w_groups["X"] > xmax_foreign)
+                    | (df_w_groups["Y"] > ymax_foreign)
+                    | (df_w_groups["Z"] > zmax_foreign)
+                ),
+                ["GROUPID"],
+            ] = (
+                df_w_groups["GROUPID"].max() + 1
+            )
+
```
Comment (olelod) [inline src/flownet/data/perforation_strategy.py:204]: For a long deviated well with all connections opening/closing at the same times the bounding box defined by (xmin, xmax, ymin, ymax, zmin, zmax) can get quite large. Can this potentially make us split connections we don't want to split? Are there better options for a bounding box? Some kind of cylindrical area along the well? I guess we can add it as some kind of issue/improvement suggestion?

**Comment by olelod** [comment]
LGTM. Just one comment above. Maybe for of an improvement suggestion?

File: src/flownet/data/perforation_strategy.py (hunk 4 above)
Comment (wouterjdb) [reply src/flownet/data/perforation_strategy.py:204]: I agree, the bounding box solution is quite poor. I have indeed also thought about fitting some spline and trying o fine out the order like that. However, then there are also cases with multi laterals that are even harder to identify with the definition used in the simulator.

If you have a better solution that is still -relatively- straightforward to implement I'd happily include it in the PR as I'm not quite satisfied with the bounding box. Suggestions are welcome 🍪 

(We would be rid of this problem immediately when we go to the real data though.)

**Comment by wouterjdb** [comment]
Effect of shutting a connection clearly visible in the prior of two runs with different perforation strategies. (🔵 = `multiple_based_on_workover`, 🔴 = `time_avg_open_location`)

image

**Comment by edubarrosTNO** [comment]
What are we looking at @wouterjdb? Water-cut? And do you know where the reference simulation lies?

**Comment by wouterjdb** [comment]
> What are we looking at @wouterjdb? Water-cut? And do you know where the reference simulation lies?

Yeah, looking at water cut. Reference is somewhere in the blue cloud.

**But**, I now also spotted some problems with the new perforation strategy that needs to be fixed. Clearly, the well doesn't open in time here (coloring as before, this is the 2nd iteration though instead of the prior):

image

**Comment by wouterjdb** [comment]
I tried to use `ec2df.compdat` to extract the connection state through time. Unfortuonately `ecl2df.compdat` doesn't parse `WELOPEN` which is used in the Norne simulation to close well connections. This leads to `ecl2df.compdat` only reading `OPEN` states (closing is never done in `COMPDAT` in Norne).

It might be an option to add parsing of `WELOPEN` in the `ecl2d`f package, as suggested by @berland

**Comment by wouterjdb** [comment]
Waiting for https://github.com/equinor/ecl2df/pull/186 to be merged.

File: CHANGELOG.md [hunk 5]
```diff
@@ -8,6 +8,7 @@ This project adheres to [Semantic Versioning](https://semver.org/).
...
 - [#220](https://github.com/equinor/flownet/pull/220) User can now provide the historical _control mode_ in the production wells and _control mode_/_target_ in the injection wells. One value for each, applied to all wells.
 - [#199](https://github.com/equinor/flownet/pull/199) Added the possibility of defining a 'base' value for all parameters related to relative permeability. This provides the opportunity to interpolate between three relative permeability models. All the 'min' values will be used to construct a pessimistic or low model, the 'base' values are used for a 'base' model, and the 'high' values are used for an optimistic or high model value. The history matching can then be done with only one (two for three phase models - oil/water and gas/oil interpolated independently) relative permeability parameter(s) per SATNUM region.
+- [#197](https://github.com/equinor/flownet/pull/197) Added opening/closing of connects based on straddling/plugging/perforations through time. This PR also adds a new perforation strategy `multiple_based_on_workovers` which models the well connections with as few connections as possible taking into account opening/closing of groups of connections.
```
Comment (anders-kiaer) [inline CHANGELOG.md:11]:
```suggestion
- [#197](https://github.com/equinor/flownet/pull/197) Added opening/closing of connections based on straddling/plugging/perforations through time. This PR also adds a new perforation strategy `multiple_based_on_workovers` which models the well connections with as few connections as possible taking into account opening/closing of groups of connections. 
```

File: setup.py [hunk 6]
```diff
@@ -6,7 +6,7 @@
 REQUIREMENTS = [
     "configsuite>=0.6",
     "cwrap>=1.6",
-    "ecl2df>=0.5",
+    "ecl2df~=0.8.1",
```
Comment (anders-kiaer) [inline setup.py:9]: Have me made an issue of why this needs to be pinned?

**Review by anders-kiaer — APPROVED** [review]
Good work. Nice test :wink:

File: setup.py (hunk 6 above)
Comment (wouterjdb) [reply setup.py:9]: We don't really need to actually... I could change it to `ecl2df>=0.8.1`

File: src/flownet/data/perforation_strategy.py (hunk 4 above)
Comment (wouterjdb) [reply src/flownet/data/perforation_strategy.py:204]: Issue created #223

**Pull request merged**